File: src/unionFind/B_1717.java
package unionFind;

import java.util.*;
import java.io.*;

public class B_1717 {
	static StringBuilder sb = new StringBuilder();
	static int[] parent;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		
		parent = new int[n + 1];
		for (int i=1;i<=n;i++) parent[i] = i;
		
		for (int i=0;i<m;i++) {
			st = new StringTokenizer(br.readLine());
			int o = Integer.parseInt(st.nextToken());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			
			if (o == 0) union(a, b); else isUnion(a, b);
		}
		
		System.out.println(sb);
	}
	
	static int find(int a) {
		if (a == parent[a]) return a;
		return parent[a] = find(parent[a]);
	}
	
	static void union(int a, int b) {
		a = find(a);
		b = find(b);
		
		if (a == b) return;
		if (a < b) parent[b] = a;
		else parent[a] = b;
	}
	
	static void isUnion(int a, int b) {
		a = find(a);
		b = find(b);
		
		if (a == b) sb.append("YES"); else sb.append("NO");
		sb.append('\n');
	}
}

File: src/kotlin/unionfind/B_1717.kt
lateinit var parent: IntArray

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    parent = IntArray(n + 1) { it }

    val sb = StringBuilder()
    repeat(m) {
        val (t, a, b) = readLine().split(" ").map { it.toInt() }
        if (t == 0) {
            union(a, b)
            return@repeat
        }
        val parentX = find(a)
        val parentY = find(b)

        sb.append(if (parentX == parentY) "YES" else "NO").append('\n')
    }
    println(sb)
}

fun union(x: Int, y: Int) {
    val parentX = find(x)
    val parentY = find(y)

    if (parentX == parentY) return
    if (parentX < parentY) {
        parent[parentX] = parentY
        return
    }
    parent[parentY] = parent[parentX]
}

fun find(x: Int): Int {
    if (parent[x] == x) return x
    parent[x] = find(parent[x])
    return parent[x]
}
